Remove Resource[Service] and Resource[Endpoints]

joamaki opened the pull request:
Switch over remaining uses of `Resource[*slim_corev1.Service]` and `Resource[*k8s.Endpoints]` over to `Table[*lb.Service]` or `Table[*lb.Frontend]` in the agent. This avoids indexing and holding onto duplicate services and endpoints and thus reduces memory usage.

Results from `github.com/joamaki/cilium-lb-load-test` which creates 5000 services with 10 backends each and measures memory usage:

#### `main` branch (`66a4e8e45187f9e25235113690e0ebf6275a81f4`)
```
$ ./run-loadtest.sh
...
RSS (kB): 291900.0 (diff 136920.0)
Reachable kB: 96813.1484375
Allocations: 6006325
Allocated kB: 339197.2578125
```

#### This PR 
```
RSS (kB): 281664.0 (diff 124964.0)
Reachable kB: 80810.9140625
Allocations: 5933907
Allocated kB: 324277.0859375
```

joamaki committed:
loadbalancer/reflectors: Reflect endpoints directly

Use the [k8s.ListerWatcherToObservable] to reflect endpoints directly
without relying on Resource[Endpoints].

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
daemon: Remove Resource[Endpoints] and endpoints watcher

EndpointsWatcher and Resource[Endpoints] is no longer needed by the cilium-agent
as usages of it have migrated to Table[Backend].

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
ipcache: Watch backends table to update api-server labels

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
loadbalancer/reflectors: Reflect services directly

Stop using Resource[*Service] and reflect services directly into the tables
with [k8s.ListerWatcherToObservable]. Drop the now unused Resource[*Service].

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
loadbalancer: Fix few test flakes

Switching to directly reflecting services and endpoint slices and not using an
informer made things converge a little bit faster and revealed couple flaky tests:

- reuse.txtar was deleting and inserting in quick succession 'test' and 'test2' services
  which lead to upsert of test and test2 back to back causing a conflict. This was because
  a prior delete of 'test2' was replaced by a future upsert and this was processed before
  upsert of 'test' due to preserving the ordering. The implementation is correct as it has
  the same semantics as normal informer + workqueue setup. Fixed the test to sync after each
  sequence of deletes and updates.

- redirectpolicy's service.txtar didn't check that frontends were reconciled causing sometimes
  the reconciler to not process the deletion and thus reusing the frontend IDs when they were
  added back.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
dial: Use Table[Frontend] for looking up address

As the Resource[*Service] is being removed refactor pkg/dial to instead
look up the ClusterIP address for a service from the frontends table.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
l2announcer: Switch to using Table[Service/Frontend]

Refactor to use Table[*Service] and Table[*Frontend] instead of the
Resource[*slim_corev1.Service].

The Table[*Service] is always updated when frontends change so we're
relying on that to trigger the processing.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki commented:
/test

joamaki commented:
/test

joamaki commented:
/test

joamaki commented:
/test

marseel commented on pkg/dial/resolver.go:
```diff
@@ -28,31 +28,40 @@ import (
+// ResourceServiceResolverCell provides a ServiceResolver instance to map DNS names
+// matching Kubernetes services to the corresponding ClusterIP address using the
+// services resource. Used by the operator.
+var ResourceServiceResolverCell = cell.Module(
```
Does something prevents us from accidentally using `ResourceServiceResolverCell` in Cilium Agent and starting additional informers?

marseel commented on pkg/l2announcer/l2announcer.go:
```diff
@@ -139,17 +139,18 @@ func NewL2Announcer(params l2AnnouncerParams) *L2Announcer {
 func (l2a *L2Announcer) run(ctx context.Context, health cell.Health) error {
 	var err error
-	l2a.svcStore, err = l2a.params.Services.Store(ctx)
+	l2a.policyStore, err = l2a.params.L2AnnouncementPolicy.Store(ctx)
```
nit: Is there a reason why order was swapped? I guess not, but just double checking. 
Before, we had services -> policy, now we have policy -> service

marseel commented on pkg/l2announcer/l2announcer.go:
```diff
@@ -139,17 +139,18 @@ func NewL2Announcer(params l2AnnouncerParams) *L2Announcer {
 
-	l2a.policyStore, err = l2a.params.L2AnnouncementPolicy.Store(ctx)
+	wtxn := l2a.params.StateDB.WriteTxn(l2a.params.Services)
+	svcChangeIter, err := l2a.params.Services.Changes(wtxn)
```
note to myself: this pushes down waiting on initialization compared to `.Store`

joamaki committed:
loadbalancer: Fix flake in hybrid-dsr.txtar

The TCP and UDP services were added at the same time which
sometimes caused the IDs to be allocated in an unexpected
order. Fix this by adding the TCP service first and waiting
for it to get reconciled and only then add the UDP service.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

marseel commented on pkg/dial/resolver_test.go:
```diff
@@ -36,73 +36,67 @@ import (
-)
-
-func TestServiceResolver(t *testing.T) {
+func TestLBServiceResolver(t *testing.T) {
```
very nice, thanks!

tklauser reviewed (approved):
operator changes look good. Didn't review for my other codeowners (thus re-requesting review from @cilium/ipcache)

gandro commented on daemon/cmd/daemon.go:
```diff
@@ -383,10 +383,6 @@ func configureDaemon(ctx context.Context, params daemonParams) error {
 	if params.DaemonConfig.DNSPolicyUnloadOnShutdown {
-		// Stop k8s watchers
-		params.Logger.Info("Stopping k8s watcher")
-		params.K8sWatcher.StopWatcher()
```
>  I doubt it was there for that reason. I think it was mostly just that tests don't leak resources.

This is not a correct assumption. The reason we stop the watcher is that we want to make sure that no new policies are ingested after we've removed all DNS rules from `params.Policy.Iterate`. Arguably this code was already kinda broken, but we shouldn't break it further and instead fix this.

Probably it would be cleaner to have some mechanism in the policy importer or repository that denies ingesting new policies, rather than stopping the K8s watch. But I don't think we can just remove this, as it would potentially break the DNS unload on shutdown mechanism.

gandro reviewed (approved):
Apparently my concerns regarding DNSUnloadOnShutdown were not correct (c.f. https://github.com/cilium/cilium/pull/42440/files#r2584707929)
